hypertrace-trace-enricher/trace-reader/src/main/java/org/hypertrace/trace/reader/attributes/DefaultValueResolver.java

```diff
@@ -72,108 +73,106 @@ private Single<LiteralValue> resolveDefinition(
+  private Optional<LiteralValue> resolveFirstValuePresent(
       ValueSource valueSource,
       AttributeMetadata attributeMetadata,
       AttributeDefinitions definitions) {
 
-    return Observable.fromIterable(definitions.getDefinitionsList())
-        .concatMapMaybe(
-            definition -> this.maybeResolveDefinition(valueSource, attributeMetadata, definition))
-        .first(LiteralValue.getDefaultInstance());
+    return definitions.getDefinitionsList().stream()
+        .map(definition -> this.resolveDefinition(valueSource, attributeMetadata, definition))
+        .filter(Optional::isPresent)
+        .findFirst()
+        .flatMap(Function.identity());
   }
 
-  private Single<LiteralValue> resolveExpression(
+  private Optional<LiteralValue> resolveExpression(
       ValueSource valueSource, ProjectionExpression expression) {
-
-    Single<AttributeProjection> projectionSingle =
-        Maybe.fromOptional(this.attributeProjectionRegistry.getProjection(expression.getOperator()))
-            .switchIfEmpty(
-                buildAndLogErrorLazily(
-                    "Unregistered projection operator: %s", expression.getOperator()));
-    Single<List<LiteralValue>> argumentsSingle =
+    Optional<AttributeProjection> maybeProjection =
+        this.attributeProjectionRegistry.getProjection(expression.getOperator());
+    if (maybeProjection.isEmpty()) {
+      if (LOGGING_LIMITER.tryAcquire()) {
+        log.error("Unregistered projection operator: {}", expression.getOperator());
+      }
+      return Optional.empty();
+    }
+    List<LiteralValue> argumentList =
         this.resolveArgumentList(valueSource, expression.getArgumentsList());
 
-    return zip(projectionSingle, argumentsSingle, AttributeProjection::project);
+    if (argumentList.isEmpty()) {
+      if (LOGGING_LIMITER.tryAcquire()) {
+        log.error(
+            "Failed to resolve argument list for expression with operator: {}",
+            expression.getOperator());
+      }
+      return Optional.empty();
+    }
+    return Optional.of(maybeProjection.get().project(argumentList));
   }
 
-  private Single<List<LiteralValue>> resolveArgumentList(
+  private List<LiteralValue> resolveArgumentList(
       ValueSource valueSource, List<Projection> arguments) {
-    return Observable.fromIterable(arguments)
-        .flatMapSingle(argument -> this.resolveProjection(valueSource, argument))
-        .collect(Collectors.toList());
-  }
-
-  private <T> Single<T> buildAndLogErrorLazily(String message, Object... args) {
-    return Single.error(
-        () -> {
-          if (LOGGING_LIMITER.tryAcquire()) {
-            log.error(String.format(message, args));
-          }
-          return new UnsupportedOperationException(String.format(message, args));
-        });
+    Stream<Optional<LiteralValue>> resolvedArguments =
+        arguments.stream().map(argument -> this.resolveProjection(valueSource, argument));
```

Comment:
If this is the only place we're using it and we don't support it return empty, why not just throw? You're eventually throwing anyway, you're just going through an extra level of indirection. If your goal is to completely avoid the throw, collect to list (use flatmap to optional stream) and just compare size before returning.